src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,216 @@
+    /// </remarks>
+    internal sealed class SemanticTokensCache
+    {
+        private const int MaxSemanticVersionPerDoc = 4;
+
+        // Nested cache mapping (URI -> (semanticVersion -> (line #s -> tokens on line)))
+        private readonly MemoryCache<DocumentUri, MemoryCache<VersionStamp, Dictionary<int, IReadOnlyList<int>>>> _cache = new();
+
+        /// <summary>
+        /// Caches tokens on a per-line basis. If the given line has already been cached for the document
+        /// and its semantic version, the line's tokens will be ignored.
+        /// </summary>
+        /// <param name="uri">The URI associated with the passed-in tokens.</param>
+        /// <param name="semanticVersion">The document's semantic version associated with the passed-in tokens.</param>
+        /// <param name="range">The range associated with the passed-in tokens.</param>
+        /// <param name="tokens">The complete set of tokens for the passed-in range.</param>
+        internal void CacheTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range range,
+            int[] tokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache))
+            {
+                documentCache = new MemoryCache<VersionStamp, Dictionary<int, IReadOnlyList<int>>>(sizeLimit: MaxSemanticVersionPerDoc);
+                _cache.Set(uri, documentCache);
+            }
+
+            if (!documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                lineToTokensDict = new Dictionary<int, IReadOnlyList<int>>();
+                documentCache.Set(semanticVersion, lineToTokensDict);
+            }
+
+            var absoluteLine = 0;
+
+            // Cache the tokens associated with each line
+            for (var tokenIndex = 0; tokenIndex < tokens.Length; tokenIndex += 5)
+            {
+                absoluteLine += tokens[tokenIndex];
+
+                // Don't cache tokens outside of the given range
+                if (absoluteLine < range.Start.Line)
+                {
+                    continue;
+                }
+
+                if (absoluteLine >= range.End.Line)
+                {
+                    break;
+                }
+
+                if (lineToTokensDict.TryGetValue(absoluteLine, out _))
+                {
+                    // This line is already cached, so we can skip it
+                    continue;
+                }
+
+                var lineTokens = new List<int>();
+
+                // Cache the line's tokens
+                while (tokenIndex < tokens.Length)
+                {
+                    lineTokens.Add(tokens[tokenIndex]);
+                    lineTokens.Add(tokens[tokenIndex + 1]);
+                    lineTokens.Add(tokens[tokenIndex + 2]);
+                    lineTokens.Add(tokens[tokenIndex + 3]);
+                    lineTokens.Add(tokens[tokenIndex + 4]);
```

Comment:
is there an addrange method we could use here instead?

Reply:
Not for Lists (as fa as I know) ☹️

Reply:
@ryanbrandenburg can you use an ImmutableArray.Builder?